simpleParser.go

```diff
@@ -0,0 +1,470 @@
+		ctx.skipAdvanceMode = false
+		return nil
+	}
+
+	// context mode transition
+	switch ctx.subCtx.currentMode {
+	case fieldMode:
+		// After the field mode, the next token *must* be an op
+		ctx.subCtx.currentMode = opMode
+	case opMode:
+		// After the op mode, the next mode should be value mode
+		ctx.subCtx.currentMode = valueMode
+	case valueMode:
+		if ctx.subCtx.oneLayerMode {
+			// One layer mode means that we should return as soon as this value is done so we can merge
+			ctx.currentTokenIndex = ctx.currentTokenIndex - 1
+			return NonErrorOneLayerDone
+		} else {
+			// After the value is finished, this subcompletion becomes a "field", so the next mode should be
+			// a op
+			ctx.subCtx.currentMode = opMode
+		}
+	default:
+		return fmt.Errorf("Not implemented yet for mode transition %v", ctx.subCtx.currentMode)
+	}
+	return nil
+}
+
+func (ctx *expressionParserContext) handleOpenParen() error {
+	// Strip the "(" from this token and into its own element
+	ctx.tokens = append(ctx.tokens, "")
+	ctx.tokens[ctx.currentTokenIndex] = strings.TrimPrefix(ctx.tokens[ctx.currentTokenIndex], "(")
+	copy(ctx.tokens[ctx.currentTokenIndex+1:], ctx.tokens[ctx.currentTokenIndex:])
+	ctx.tokens[ctx.currentTokenIndex] = "("
+	ctx.parenDepth++
+	// for open paren, internally advance so the next getToken will not get a paren
+	ctx.currentTokenIndex = ctx.currentTokenIndex + 1
+	return nil
+}
+
+func (ctx *expressionParserContext) handleCloseParen() error {
+	// Strip the ")" from the end of this token and insert into its own element
+	for strings.HasSuffix(ctx.tokens[ctx.currentTokenIndex], ")") {
+		ctx.tokens = append(ctx.tokens, "")
+		copy(ctx.tokens[ctx.currentTokenIndex+1:], ctx.tokens[ctx.currentTokenIndex:])
+		ctx.tokens[ctx.currentTokenIndex] = strings.TrimSuffix(ctx.tokens[ctx.currentTokenIndex], ")")
+		ctx.tokens[ctx.currentTokenIndex+1] = ")"
+	}
+	return nil
+}
+
+func (ctx *expressionParserContext) handleCloseParenBookKeeping() error {
+	if ctx.parenDepth == 0 {
+		return ErrorParenMismatch
+	}
+	ctx.parenDepth--
+	return nil
+}
+
+func (ctx *expressionParserContext) getCurrentToken() (string, ParseTokenType, error) {
+	if ctx.currentTokenIndex >= len(ctx.tokens) {
+		return "", TokenTypeInvalid, ErrorNoMoreTokens
+	}
+	token := ctx.tokens[ctx.currentTokenIndex]
+
+	// Values by def should be enclosed within single quotes
+	valueRegex := regexp.MustCompile(`^\'.*\'$`)
+	// Or Values can be int or floats by themselves (w/o alpha char)
+	valueNumRegex := regexp.MustCompile(`^(-?)(0|([1-9][0-9]*))(\\.[0-9]+)?$`)
+
+	if tokenIsOpType(token) {
+		return token, TokenTypeOperator, nil
+	} else if valueRegex.MatchString(token) {
+		return token, TokenTypeValue, nil
+	} else if valueNumRegex.MatchString(token) {
+		return token, TokenTypeValue, nil
+	} else if token == "true" {
+		return token, TokenTypeTrue, nil
+	} else if token == "false" {
+		return token, TokenTypeFalse, nil
+	} else if token != "(" && strings.HasPrefix(token, "(") {
```

Comment:
for completeness, should we handle token == "("?
if so, the code may be more symmetric, i.e., after ctx.handleOpenParen() in this block, we can simply return ctx.getCurrentToken()?